File: problems/38-advanced-graph-combinations/001-critical-connections/java/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    private List<Integer>[] graph;
    private int[] disc;
    private int[] low;
    private int time;
    private List<List<Integer>> bridges;

    public List<List<Integer>> criticalConnections(int n, List<List<Integer>> connections) {
        graph = new ArrayList[n];
        for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
        for (List<Integer> edge : connections) {
            int u = edge.get(0), v = edge.get(1);
            graph[u].add(v);
            graph[v].add(u);
        }

        disc = new int[n];
        low = new int[n];
        bridges = new ArrayList<>();
        time = 1;

        for (int i = 0; i < n; i++) {
            if (disc[i] == 0) dfs(i, -1);
        }
        return bridges;
    }

    private void dfs(int u, int parent) {
        disc[u] = low[u] = time++;

        for (int v : graph[u]) {
            if (v == parent) continue;
            if (disc[v] == 0) {
                dfs(v, u);
                low[u] = Math.min(low[u], low[v]);
                if (low[v] > disc[u]) bridges.add(Arrays.asList(u, v));
            } else {
                low[u] = Math.min(low[u], disc[v]);
            }
        }
    }
}

File: problems/38-advanced-graph-combinations/001-critical-connections/python/solution.py
def critical_connections(n: int, connections: list[list[int]]) -> list[list[int]]:
    graph = [[] for _ in range(n)]
    for u, v in connections:
        graph[u].append(v)
        graph[v].append(u)

    discovery = [-1] * n
    low = [-1] * n
    bridges = []
    time = 0

    def dfs(u: int, parent: int) -> None:
        nonlocal time
        discovery[u] = low[u] = time
        time += 1

        for v in graph[u]:
            if v == parent:
                continue
            if discovery[v] == -1:
                dfs(v, u)
                low[u] = min(low[u], low[v])
                if low[v] > discovery[u]:
                    bridges.append([u, v])
            else:
                low[u] = min(low[u], discovery[v])

    for node in range(n):
        if discovery[node] == -1:
            dfs(node, -1)

    return bridges
